NGコマンドの文字列処理の方法を変更

kurotoris opened the pull request:
前のPRの停止ありがとうございます
今回は速度はほんの少し落ちてる気がする程度で済んでいます
あと、NGコマンドの文字処理にsearch()を使用することで正規表現に対応させました

変更点としては下記のようになります
NGListManagerの最初の方に　ngCommandAll（ALL用フラグ）　ngCommandList（NGコマンドリスト）　の宣言をしました
refreshNgMap　をかなり変更しました
NGコマンドに"ALL"の文字列があるかどうか確認し、あった場合は　ngCommandAll　にtrueを設定する
NGコマンド文字列を　ngCommandList　に放り込む処理をする
これに伴いNGコマンド文字列は　NgMap　には入れていません
isNgCommand　は　ngCommandAll　にtrueが設定されている場合には"184"などのコマンドを含むかの判断をさせる
"184"コマンドは"DEVICE:3DS"など類似コマンドもあるので、受け取ったコマンドを切り分けてチェックをかけている
その処理が終了した後　ngCommandList　に入っているNGコマンドコマンド文字列とコマンド文字列を正規表現で比較し処理する
正規表現での検索に対応してる事を動作確認済み

という感じに変更しました
大きな変更点の部分は、一応ですが修正前の処理をコメント文にして残すようにしました
前回と同じように、私の変更点には日付を横に書いてあります

処理自体は、処理するコメント数が１万以上まで多くなると、微妙に差が出てる気がしますが
元のプログラムと１秒あるかないかの差の上に、全体的な動画読み込みの中に埋もれてしまうので
今回の処理は問題ないと思います

kurotoris committed:
NGコマンドの処理方法を変更しました

NGコマンドの処理方法を変更しました

NGListManagerの最初の方に　ngCommandAll（ALL用フラグ）　ngCommandList（NGコマンドリスト）　の宣言をしました

refreshNgMap　をかなり変更しました
NGコマンドに"ALL"の文字列があるかどうか確認し、あった場合は　ngCommandAll　にtrueを設定する
NGコマンド文字列を　ngCommandList　に放り込む処理をする
これに伴いNGコマンド文字列は　NgMap　には入れていません

isNgCommand　は　ngCommandAll　がある場合には"184"などのコマンドを含むかの判断をさせる
"184"コマンドは"DEVICE:3DS"など類似コマンドもあるので、受け取ったコマンドを切り分けてチェックをかけている
その処理が終了した後　ngCommandList　に入っている文字列と渡された文字列を正規表現で比較し
一致すればtrueで送り返すという処理をする
正規表現での検索に対応してる事を動作確認済み

修正前の処理は一応残すようにした
私の変更点には日付を横に書いてあります

処理自体は、処理するコメント数が万単位まで多くなると、微妙に差が出る
元のプログラムと１秒あるかないかの差なので、数万コメントとかでなければ大丈夫だと思う

SSW-SCIENTIFIC commented on src/org/mineap/nndd/player/NGListManager.as:
```diff
@@ -178,40 +208,50 @@ package org.mineap.nndd.player
+				if(Command2.match(ngCommand)){
+					return true;
+				}
+			}
```
正規表現より`Command2.indexOf(ngCommand) != -1`で判定した方がよいと思います(正規表現は非常に遅い)。
ただ微妙なところもある(コマンドの一部を拾ってしまう可能性がある)ので正規表現とするなら`new RegExp("\\b" + ngCommand + "\\b", "i")`をパターンとして使った方がいいでしょう(処理速度と正確性のトレードオフです)。

ただNgCommandListを本当にAll ofで処理するのならば`var commands = Command2.split(" ")`及び`var ngCommands = ngCommand.split(" ")`として`in commands`と`in ngCommands`の2重ループを回すのが適切かなと思います。
`ngCommands.length == 1`の場合だけ上記の`indexOf`による判定にすれば, 実際のNG登録で2重ループで判定しなければならないケース(複数コマンドのAll of 判定)は殆どないはずなので, 速度的な影響も少ない気がします。

kurotoris replied:
ＮＧコマンドの被りでueとbl"ue”が被ってる事が分かりました
ちょっと解決方法考えて、出来たら修正かけます

SSW-SCIENTIFIC replied:
```actionscript
var CommandList = Command2.split(" ");
var ngCommandList = ngCommand.split(" ");
var result = true;

for each(var command in CommandList) {
    if (ngCommandList.indexOf(command) == -1) {
        result = false;
    }
}
```

試してませんがこんなのはどうでしょうか。

kurotoris committed:
NGコマンド、正規表現対応（"blue"と"ue"が同時に検出される事を修正）

NGコマンドで"ue"コマンドをNGに入れると"blue"もNGに引っかかるので
if(Command2.match(ngCommand))　を
if(Command2.match("\\b" + ngCommand + "\\b"))　に変更
"blue"と"ue"、"red"と"red2"を区別してチェックする事を確認
ただ"DEVICE:3DS"と"3ds"を区別せずにNGにする状態になった（"3ds"と入れると"DEVICE:3DS"はNGになる）
"DEVICE:WIIU"などは引っかからないので問題無いと思う（"DEVICE"を入れると"3ds"と"WIIU"がNGになる）

コメントの処理時間については　Comments.as　の
ddCommentToArrayCollection　に一番最初と一番最後の位置にタイム測定用のコマンドを入れて測定してみました
            var old_time = getTimer();
　　　　　＜途中の処理＞
            var time = getTimer() - old_time;	// 実行後に測定(ms単位）
            trace("経過時間:" + time);		// 差
対象は　けものフレンズ1話目（手持ちで一番コメントが多かった）
デバッグモードで実行、一度動画を読み込み、再生を停止させてからNGコマンド"red"を追加する時の処理時間です
ID、単語などのNGは同じ、NGコマンドは"red"のみを新規で追加した場合
フォーク前のプログラム
[trace] commentCount:32129
[trace] maxCommentCount:144000
[trace] 経過時間:6865
[trace] Comments:投稿者コメントロード完了
[trace] Comments.loadCommentByXML:コメントをXMLから配列に格納
[trace] commentCount:0
[trace] maxCommentCount:9600

今回のプログラムでNGコマンド正規表現（他はほぼ同一、コメント数だけ多少増加なので省略）
[trace] 経過時間:6872

NGコマンド１０単語入れておいて"red"を追加した場合
（"small""big""medium""ue""shita""red2"など正規表現的ではない１０単語です）
フォーク前のプログラム
[trace] 経過時間:6824

今回のプログラムでNGコマンド正規表現
[trace] 経過時間:7331
（参考、同じ条件で、コメントが半分程度の動画だと）
[trace] commentCount:14953
[trace] maxCommentCount:144000
[trace] 経過時間:3051

という感じです
NGコマンドとして登録してある数が増えると処理が重くなっていくのですが
うちでは0.5s程度しか変わらないので、視聴の時の処理自体はそこまで影響ないのではないかと思います
いかがでしょうか？

参考にいただいたコードについては
Command2.indexOf(ngCommand) != -1　こちらは"blue"と"ue"の区別をする方法が思いつきませんでした
もう一つは、動かしてみたのですが、上手く動かなくて断念しました

SSW-SCIENTIFIC reviewed (changes requested):
修正提案
----------
`ue big`と`big ue`とで同じ結果が得られないのはやはり残念なので, 248-255行目を以下のように変更することを提案します。

```actionscript
var commands: Array = Command2.split(" ");
for each(var ngCommand:String in ngCommandList){
    var ngCommands: Array = ngCommand.split(" ");
    var matched: Boolean = true;

    for each (var ngItem: String in ngCommands) {
        if (commands.indexOf(ngItem) === -1) {
            matched = false;
            break;
        }
    }

    if (matched) {
        return true;
    }
}

return false;
```

`ngCommandList`に対してはAny ofで処理を行い, その中の各`ngCommand`は`split`で分割した上で`Command2`に全てが含まれるかを検査します。

処理速度
---------
sm116203でテストしました。
上記修正後, 200コメント/分の表示設定で`ue red`, `big red`などの処理時間は概ね100ms未満でフィルタリング出来ています(件数が増える場合は線型増加だと思われる)
修正前の正規表現を利用したコードでは, 内容にも依りますが100-200ms程度の処理時間であったため, 正規表現よりも`indexOf`による判定の方がよいと判断します。

kurotoris committed:
NGコマンド機能の処理方法変更

NGコマンド処理の方法を提案の通りに変更
split'　が他の処理と被っている場所があるため、修正
正規表現で処理していたものを　indexOf　で処理する状態に変更

引用
ngCommandListに対してはAny ofで処理を行い, その中の各ngCommandはsplitで分割した上でCommand2に全てが含まれるかを検査します。
引用終わり

kurotoris commented:
提案のあった通りに書き換えました
一部、元から　split　処理をしている場所があるので
そちらで　split　処理をする状態に書き直しました
こちらでもNG処理のチェックをしましたが、問題ありませんでした

色々と教えていただいたり、お手間を取らせて申し訳ないです
ありがとうございました

SSW-SCIENTIFIC reviewed (approved):
お疲れさまでした。マージします。
